File: algorithm/leetcodeObjectiveC/wiggle-sort.m
// 280. Wiggle Sort
// Difficulty: Medium

// Given an unsorted array nums, reorder it such that nums[0] < nums[1] > nums[2] < nums[3]....

// Example:
// (1) Given nums = [1, 5, 1, 1, 6, 4], one possible answer is [1, 4, 1, 5, 1, 6].
// (2) Given nums = [1, 3, 2, 2, 3, 1], one possible answer is [2, 3, 1, 3, 1, 2].

// Note:
// You may assume all input has valid answer.

// Follow Up:
// Can you do it in O(n) time and/or in-place with O(1) extra space?

// Time:  O(n)
// Space: O(1)

#import <Foundation/Foundation.h>

void wiggleSort(NSMutableArray* nums) {
  for (int i = 1; i < nums.count; ++i) {
    if (((i % 2) && [nums[i] isLessThan:nums[i - 1]]) ||
        (!(i % 2) && [nums[i] isGreaterThan:nums[i - 1]])) {
      [nums exchangeObjectAtIndex:i withObjectAtIndex:i - 1];
    }
  }
}

File: data-structure-and-algorithm/typescript/leetcode/src/wiggle-sort.ts
/*
Given an unsorted array nums, reorder it in-place such that nums[0] <= nums[1] >= nums[2] <= nums[3]....

Example:

Input: nums = [3,5,2,1,6,4]
Output: One possible answer is [3,5,1,6,2,4]
*/

function wiggleSort(nums: number[]): void {
  for (let i = 1; i < nums.length; ++i) {
    if (
      (i % 2 && nums[i] > nums[i - 1]) ||
      (!(i % 2) && nums[i] < nums[i - 1])
    ) {
      swap(i, i - 1);
    }
  }

  function swap(i: number, j: number): void {
    const temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}
